### PR TITLE
fix: update failure in requeue

### PR DESCRIPTION
**Pull request by kerthcet**
Signed-off-by: kerthcet <kerthcet@gmail.com>

#### What type of PR is this?
/kind bug

#### Which issue(s) this PR fixes:

Part of https://github.com/kubernetes-sigs/kueue/issues/241

### CODE REVIEW

File: pkg/queue/manager.go [hunk 1]
```diff
@@ -300,13 +300,14 @@ func (m *Manager) RequeueWorkload(ctx context.Context, info *workload.Info, imme
 		return false
 	}
 
-	q.AddIfNotPresent(info)
+	newInfo := workload.NewInfo(&w)
```
Comment (alculquicondor) [inline pkg/queue/manager.go:303]: Can you explain why you think this solves the problem?

If there was a change in the object, the event would update the object, and `AddIfNotPresent` would prevent it to be added again.

Comment (kerthcet) [reply pkg/queue/manager.go:303]: Refer to the log, we are always failing with update operation, this must be the reason that we stored the old version workload. So we should ensure to store the latest workload. The reason might be that the failed api update with the old version  requeue earlier than the newest one, then we stored the old ones. We can also find some clues from the log:
image

Comment (alculquicondor) [reply pkg/queue/manager.go:303]: But that screenshot suggests that it wasn't actually added. So I'm not sure if this solves the issue or maybe it obscures a different one.

What's the link for this failure? I want to look at the entire logs.

Comment (kerthcet) [reply pkg/queue/manager.go:303]: Here's the link https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_kueue/227/pull-kueue-test-integration-main/1520594730851766272/build-log.txt

Comment (kerthcet) [reply pkg/queue/manager.go:303]: image
From the context, we know that the `assume-1` is return from the old version, it requeued successfully, the `assume-2` is from the new version, it requeued failed for old workload has already requeued.

File: pkg/queue/manager.go [hunk 2]
```diff
@@ -300,13 +300,14 @@ func (m *Manager) RequeueWorkload(ctx context.Context, info *workload.Info, imme
 		return false
```
Comment (alculquicondor) [inline pkg/queue/manager.go:300]: actually, a new version might be on a different queue (because the user might have changed it).

let's do the get before obtaining the queue from `m.queues`.

File: pkg/queue/manager.go [hunk 3]
```diff
@@ -300,13 +300,14 @@ func (m *Manager) RequeueWorkload(ctx context.Context, info *workload.Info, imme
...
-	q.AddIfNotPresent(info)
+	newInfo := workload.NewInfo(&w)
+	q.AddIfNotPresent(newInfo)
 	cq := m.clusterQueues[q.ClusterQueue]
 	if cq == nil {
 		return false
 	}
 
-	added := cq.RequeueIfNotPresent(info, immediate)
+	added := cq.RequeueIfNotPresent(newInfo, immediate)
```
Comment (alculquicondor) [inline pkg/queue/manager.go:310]: Should we change this for just `Requeue` instead of `RequeueIfNotPresent`?

It might not be necessary, as one of the routines trying to requeue will have the chance to push the newest version.

File: pkg/queue/manager_test.go [hunk 4]
```diff
@@ -710,6 +710,10 @@ func TestHeadsAsync(t *testing.T) {
 	now := time.Now().Truncate(time.Second)
 	cq := utiltesting.MakeClusterQueue("fooCq").Obj()
 	wl := kueue.Workload{
+		TypeMeta: metav1.TypeMeta{
```
Comment (alculquicondor) [inline pkg/queue/manager_test.go:713]: why is this necessary?

**Review by alculquicondor — COMMENTED** [review]
Can you add unit tests?

File: pkg/queue/manager.go (hunk 2 above)
Comment (kerthcet) [reply pkg/queue/manager.go:300]: good catch.

File: pkg/queue/manager_test.go (hunk 4 above)
Comment (kerthcet) [reply pkg/queue/manager_test.go:713]: We get workloads from api server now which contains these necessary fields.

File: pkg/queue/manager.go (hunk 3 above)
Comment (kerthcet) [reply pkg/queue/manager.go:310]: sg.

File: pkg/queue/manager_test.go (hunk 4 above)
Comment (alculquicondor) [reply pkg/queue/manager_test.go:716]: Maybe use `cmpopts.IgnoreTypes` instead.

File: pkg/queue/manager.go [hunk 5]
```diff
@@ -293,25 +293,26 @@ func (m *Manager) RequeueWorkload(ctx context.Context, info *workload.Info, imme
 	m.Lock()
 	defer m.Unlock()
 
-	q := m.queues[queueKeyForWorkload(info.Obj)]
-	if q == nil {
-		return false
-	}
-
 	var w kueue.Workload
+	// Always to get the newest workload in case of requeuing the out of date obj.
```
Comment (alculquicondor) [inline pkg/queue/manager.go:297]:
```suggestion
	// Always get the newest workload to avoid requeuing an out-of-date obj.
```

File: pkg/queue/manager_test.go [hunk 6]
```diff
@@ -708,21 +708,37 @@ func TestHeadsAsync(t *testing.T) {
...
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{Name: "bar"},
+			Spec: kueue.QueueSpec{
+				ClusterQueue: "barCq",
+			},
 		},
 	}
-	wantHeads := []workload.Info{
+	defaultHeads := []workload.Info{
```
Comment (alculquicondor) [inline pkg/queue/manager_test.go:741]: get rid of this variable and put it in the test table when necessary.

Comment (alculquicondor) [reply pkg/queue/manager_test.go:741]: No, I just mean to get rid of this variable and copy paste the list of workloads into the test table.

File: pkg/queue/manager.go [hunk 7]
```diff
@@ -293,25 +293,26 @@ func (m *Manager) RequeueWorkload(ctx context.Context, info *workload.Info, imme
...
+		return false
+	}
+	q.AddOrUpdate(&w)
 	q.reportPendingWorkloads()
 	cq := m.clusterQueues[q.ClusterQueue]
 	if cq == nil {
 		return false
 	}
 
+	info = workload.NewInfo(&w)
```
Comment (alculquicondor) [inline pkg/queue/manager.go:315]: Once we do #162, we probably want to make this `info.Update(&w)` or similar.

Can you leave a `TODO(#162)`?

File: pkg/queue/manager_test.go [hunk 8]
```diff
@@ -700,6 +701,8 @@ func TestHeads(t *testing.T) {
 	}
 }
 
+var ignoreTypeMeta = cmpopts.IgnoreFields(metav1.TypeMeta{}, "Kind", "APIVersion")
```
Comment (alculquicondor) [inline pkg/queue/manager_test.go:704]:
```suggestion
var ignoreTypeMeta = cmpopts.IgnoreType(metav1.TypeMeta{})
```

File: pkg/queue/manager.go (hunk 7 above)
Comment (kerthcet) [reply pkg/queue/manager.go:316]: Done.

Comment (kerthcet) [reply pkg/queue/manager.go:316]: I'll keep in mind then. Thanks for pointing out.

**Comment by kerthcet** [comment]
The https://github.com/kubernetes-sigs/kueue/pull/248 bug occurs, I'll take a spike.

**Comment by alculquicondor** [comment]
/hold

could it be related to this PR? Have you seen the same error for other PRs?

**Comment by kerthcet** [comment]
> 

I think it's not related. I used to meet this error for multi times in my local dev env. I think we can ignore the error in the pr.

**Review by alculquicondor — COMMENTED** [review]
/lgtm
/hold cancel
